internal/agent/device/osimage.go

```diff
@@ -0,0 +1,90 @@
+package device
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/device/status"
+	"github.com/flightctl/flightctl/internal/container"
+	"github.com/flightctl/flightctl/pkg/executer"
+	"github.com/sirupsen/logrus"
+	"k8s.io/klog/v2"
+)
+
+const (
+	RebootingReason       = "Rebooting"
+	OsImageDegradedReason = "OSImageControllerDegraded"
+)
+
+type OSImageController struct {
+	bootc         *container.BootcCmd
+	statusManager status.Manager
+	log           *logrus.Logger
+	logPrefix     string
+}
+
+func NewOSImageController(
+	executor executer.Executer,
+	statusManager status.Manager,
+	log *logrus.Logger,
+	logPrefix string,
+) *OSImageController {
+	return &OSImageController{
+		bootc:     container.NewBootcCmd(executor),
+		statusManager: statusManager,
+		log:       log,
+		logPrefix: logPrefix,
+	}
+}
+
+func (c *OSImageController) Sync(ctx context.Context, desired *v1alpha1.RenderedDeviceSpec) error {
+	klog.V(4).Infof("%s syncing device image", c.logPrefix)
+	defer klog.V(4).Infof("%s finished syncing device image", c.logPrefix)
+
+	err := c.ensureImage(ctx, desired)
+	if err != nil {
+		if err := c.statusManager.UpdateConditionError(ctx, OsImageDegradedReason, err); err != nil {
+			klog.Errorf("Failed to update condition: %v", err)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (c *OSImageController) ensureImage(ctx context.Context, desired *v1alpha1.RenderedDeviceSpec) error {
+	if desired.Os == nil {
+		klog.V(4).Infof("%s device os image is nil", c.logPrefix)
+		return nil
+	}
+
+	host, err := c.bootc.Status(ctx)
+	if err != nil {
+		return err
+	}
+
+	if host.Status.Booted.Image.Image.Image == desired.Os.Image {
+		klog.V(4).Infof("Host is already running os image %s", desired.Os.Image)
+		return nil
+	}
+
+	image := desired.Os.Image
+	c.log.Infof("Switching to os image %s", image)
+	os := *desired.Os
+	if err := c.bootc.Switch(ctx, os.Image); err != nil {
+		return err
+	}
+
+	// Update the status to progressing
+	if err := c.statusManager.UpdateCondition(ctx, v1alpha1.DeviceProgressing, v1alpha1.ConditionStatusTrue, RebootingReason, fmt.Sprintf("Rebooting into new os image: %s", image)); err != nil {
+		return err
+	}
+
+	c.log.Infof("Os image switch complete rebooting into new image")
+	if err := c.bootc.Apply(ctx); err != nil {
+		return err
```

Comment:
Any cleanup needed?

Reply:
the previous commands do leave a copy of the previous image. It could be potentially feasabile to clean that up once we are stable but that would be a future optimization. As soon as Apply is triggered the device will reboot.

Reply:
I guess garbage is fine as long as (1) it doesn't interfere with subsequent attempts and (2) it eventually gets cleaned up so the filesystem doesn't fill up.  It can be a future task.

Reply:
the image pull itself could take the entire device workload down so lots of thought needs to happen here.

Reply:
But we need to start somewhere